mithril-stm/src/membership_commitment/merkle_tree/tree.rs

```diff
@@ -25,11 +26,13 @@ pub struct MerkleTree<D: Digest> {
     n: usize,
     /// Phantom type to link the tree with its hasher
     hasher: PhantomData<D>,
+    /// Phantom type to link the tree with its leafe
+    leaves: PhantomData<L>,
 }
 
-impl<D: Digest + FixedOutput> MerkleTree<D> {
+impl<D: Digest + FixedOutput, L: MerkleTreeLeaf> MerkleTree<D, L> {
     /// Provided a non-empty list of leaves, `create` generates its corresponding `MerkleTree`.
-    pub(crate) fn new(leaves: &[MerkleTreeLeaf]) -> MerkleTree<D> {
+    pub(crate) fn new(leaves: &[MerkleTreeConcatenationLeaf]) -> MerkleTree<D, L> {
```

Comment:
The Merkle tree should not depend on any implementation of the `MerkleTreeLeaf` trait (except in the tests):
```suggestion
    pub(crate) fn new(leaves: &[L]) -> MerkleTree<D, L> {
```